SofascoreAcademy/SofascoreAcademy/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  BaseApiClient.swift
+//  SofascoreAcademy
+//
+//  Created by Akademija on 01.05.2024..
+//
+
+import Foundation
+
+class BaseApiClient {
+    
+    func sendRequest(url: URL?) async -> Result<Data, ApiError> {
+        guard let url = url else {
+            return .failure(.wrongRequest)
+        }
+        
+        do {
+            let (data, response) = try await URLSession.shared.data(for: URLRequest(url: url))
+            
+            guard let response = response as? HTTPURLResponse else {
+                return .failure(.notResults)
+            }
+            
+            switch response.statusCode {
+            case 200...299:
+                return .success(data)
+            case 404:
+                return .failure(.notFound)
+            default:
+                return .failure(.serverError(code: response.statusCode))
+            }
+        } catch {
+            return .failure(.unknown)
+        }
+    }
+    
+    func sendRequestAndParse<T: Decodable>(url: URL?, responseModel: T.Type) async -> Result<T, ApiError> {
+        let result: Result<Data, ApiError> = await sendRequest(url: url)
+        
+        do {
+            let data: Data = try result.get()
+            guard let decodedResponse = try? JSONDecoder().decode(responseModel, from: data) else {
+                return .failure(.parsingError)
+            }
+            return .success(decodedResponse)
+        } catch let error {
+            return .failure(error as? ApiError ?? .unknown)
+        }
+    }
+}
+
+extension BaseApiClient {
+    
+    private func urlWithPath(_ path: String) -> URL? {
+        var components = URLComponents()
+        components.scheme = "https"
+        components.host = "academy-backend.sofascore.dev"
+        components.path = path
+        return components.url
+    }
+    
+    func eventsForSportAndDateUrl(slug: String, date: String) -> URL? {
+        return urlWithPath("/sport/\(slug)/events/\(date)")
+    }
+    
+    func teamLogoUrl(id: Int) -> URL? {
+        return urlWithPath("/team/\(id)/image")
+    }
+    
+    func tournamentLogoUrl(id: Int) -> URL? {
+        return urlWithPath("/tournament/\(id)/image")
```

Comment:
Dobar put, ali kad budeš imala ovakih 100 funckija, onda treba malo organizirati... Npr extension ApiClient+Event, ApiClient+Team...

Reply:
Hm, mislila sam da je urednije imati sve urlove na jednom mjestu, ali vjv ima smisla imati u datoteci gdje se izvrsava funkcija koja radi taj url request. Onda mi vjv ni ne treba posebna funkcija koja stvara url za odredeni poziv, nego mogu samo to staviti na pocetku funkcije kojoj treba? A onda te funkcije koje su sada u ApiClient, razvrstam u ApiClient+nesto fileove

Reply:
A trenutno kad ih nemaš puno i kad su get lako se snađeš... ALi kad ih proširiš da primaju razne Http metode, queryItems, header parametre... onda postane malo neuredno.

Da, recimo na League sreenu pozivaš 5,6 enpointa koje kreću s pathom /tournament/..., onda ima smisla to organizirati u ApiClient+Tournament... 

Ako sam dobro skužio pitanje, možeš ostaviti taj "/tournament/" početni na svim mjestima npr.
"/tournament/\(id)/image"
"/tournament/\(id)/event/"
....